chain/network/src/peer/peer_actor.rs

```diff
@@ -661,6 +661,11 @@ impl Actor for PeerActor {
 
     fn stopped(&mut self, _ctx: &mut Self::Context) {
         Arbiter::current().stop();
+        let metric_label = &self.peer_addr.to_string();
+        // Garbage collect our metrics, ignoring the case where the metric is
+        // already deleted, to avoid races when a peer tries to reconnect.
+        let _ = metrics::PEER_DATA_READ_BUFFER_SIZE.remove_label_values(&[metric_label]);
+        let _ = metrics::PEER_DATA_WRITE_BUFFER_SIZE.remove_label_values(&[metric_label]);
```

Comment:
wouldn't it be better to put this into Codec drop() as well?

Reply:
especially since gauges are not allocated by the PeerActor

Reply:
ok, I see now that multiple codecs will hold the same gauge. How about a Guard for the Gauge itself?
Then it could be shared via Arc<Guard>.

Reply:
prometheus gauge doesn't know its parent gauge-vec, and I don't want to teach codec about specific metrics. 

We *could* add a RAII guard here, but I'd rather not:

* as a meta rule, I am not a huge fan of downstream library extensions (though, tbh I can't say I enjoy prometheus' APIs). 
* API-wise, I think its important to keep a metric just an `Arc<AtomicI64>` -- metric should be low-cost
* practically, we use two Codec objects for write buffer, so we'd have to use an extra arc and indirection there.
* fundamentally doing this in codec is quite a hack -- our infra for reading and writing data from the socket seems way more convoluted that it needs to be (especially given that it isn't particualry effective at preventing buffer growth). So I'd rather do a minimal thing here rather than invest into hack. I have a feeling that once we unfork the infra, the implementation would become much simpler

All that being said, would be happy to RAII this if you think that'd be better :)